### PR TITLE
Add experimental nio-transport plugin with http

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit adds an nio-transport plugin. It builds upon the nio
transport work in test:framework. However, it adds http functionality.
The http functionality is added by adapting a netty embedded channel.
For now, the bulk of the nio networking and tcp transport will remain in
test:framework. This allows it to be used as a tcp transport for
internal clusters.

This plugin is unique in that it depends on test:framework for the nio
work and the netty4 transport module for http (cors, pipeling, etc)
handlers.

### CODE REVIEW

**Comment by tbrooks8** [comment]
There are a couple things to note about this:

1. I'm setting `dependencyLicenses.enabled = false` right now in the build.gradle file. I'm not sure if this is the correct choice. It's just that, as we depend on test:framework, it is pulling in a bunch of dependencies that I do not actually need / use. And I was having trouble working out all of those licenses right now.
2. I duplicated `Netty4HttpClient` in this plugin's test package. At some point that will be replaced by a version that is driven by the niobium's work.
3. In the `NioPlugin` additionalSettings method I am doing the following:

```
    @Override
    public Settings additionalSettings() {
        final Settings.Builder settingsBuilder = Settings.builder();
        settingsBuilder.put(NetworkModule.TRANSPORT_TYPE_KEY, NIO_TRANSPORT_NAME);
        settingsBuilder.put(NetworkModule.HTTP_TYPE_KEY, NIO_HTTP_TRANSPORT_NAME);
        return settingsBuilder.build();
    }
```

I'm not sure if that is the correct approach. But without this the else the external cluster kept using the netty transport in tests.

**Review by s1monw — COMMENTED** [review]
I did a first pass and I think it looks pretty good as a first iteration.

**Comment by nik9000** [comment]
> It's just that, as we depend on test:framework

Is it a good idea for a production plugin to depend on the test framework? Even if it is experimental I think that is a thing we should avoid. I suppose if we have a bootstrap check to make sure no one uses it in production then we can depend on the test framework until we drop it.

File: core/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java [hunk 1]
```diff
@@ -245,6 +245,11 @@ public void apply(Settings value, Settings current, Settings previous) {
                     HttpTransportSettings.SETTING_HTTP_MAX_HEADER_SIZE,
                     HttpTransportSettings.SETTING_HTTP_MAX_INITIAL_LINE_LENGTH,
                     HttpTransportSettings.SETTING_HTTP_RESET_COOKIES,
+                    HttpTransportSettings.SETTING_HTTP_TCP_NO_DELAY,
```
Comment (tbrooks8) [inline core/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java]: Opened here: https://github.com/elastic/elasticsearch/pull/26310

File: plugins/transport-nio/build.gradle [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+dependencyLicenses.enabled = false
+
+compileJava.options.compilerArgs << "-Xlint:-try"
+compileTestJava.options.compilerArgs << "-Xlint:-rawtypes,-unchecked"
+
+dependencies {
+  compile project(path: ':modules:transport-netty4', configuration: 'runtime')
+  compile project(path: ':test:framework', configuration: 'runtime')
```
Comment (rjernst) [inline plugins/transport-nio/build.gradle]: What is it in the test framework needed here?

Comment (tbrooks8) [inline plugins/transport-nio/build.gradle]: The nio transport classes are in test:framework. Selectors, channels, etc.

File: plugins/transport-nio/src/main/java/org/elasticsearch/transport/nio/NioPlugin.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.Transport;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class NioPlugin extends Plugin implements NetworkPlugin {
```
Comment (tbrooks8) [inline plugins/transport-nio/src/main/java/org/elasticsearch/transport/nio/NioPlugin.java]: I added a check.

File: plugins/transport-nio/build.gradle (hunk 2 above)
Comment (rjernst) [inline plugins/transport-nio/build.gradle]: I thought they were going to move into this plugin?

Comment (tbrooks8) [inline plugins/transport-nio/build.gradle]: When I spoke to @s1monw about this last week, his preference was to leave the tcp version of the nio transport in test:framework for now for for testing.

**Comment by s1monw** [comment]
@tbrooks8 are we ready to move here? Shall I take another look?

**Comment by tbrooks8** [comment]
@s1monw yes. I made changes based on all your feedback that was specific for nio. I can submit a follow up PR for the changes that apply to both this and the netty module.

**Comment by tbrooks8** [comment]
This PR has been opened for a while and I think is ready for more reviews as I have address @s1monw concerns. Although I do have a concern about how this will work. 

@jasontedor I guess that when this gets merged I will put the http plugin part on 7.0 and back port the nio transport changes only to 6.X? As I think you said you wanted the plugin part on 7.0 only?

**Comment by tbrooks8** [comment]
I don't need any reviews on this PR currently. There has been a number of changes to the various nio classes. This PR will need some conflicts resolved before it is ready to be reviewed again.

**Comment by tbrooks8** [comment]
I am closing this PR. It is well out of date with updates to nio. I will eventually use some of this work to submit a separate PR for http.

**Pull request closed without merging**